File: client/src/app/core/models/view-models/import-session.ts
export enum ImportStage {
  uploaded = 0,
  inspected = 1,
  mapped = 2,
  previewed = 3,
  committing = 4,
  committed = 5,
  failed = 6,
  undone = 7,
  abandoned = 8,
}

export enum ImportDiagnosticSeverity {
  error = 0,
  warning = 1,
  info = 2,
}

export enum ImportRowAction {
  create = 0,
  update = 1,
  skip = 2,
  error = 3,
}

export interface ImportSessionViewModel {
  publicId: string;
  stage: ImportStage;
  sourceKind: number;
  vendorProfile: number;
  originalName: string;
  sizeBytes: number;
  targetRecordType: string;
  targetProjectKey?: string;
  targetBoardIdentifier?: string;
  progressPercent: number;
  progressMessage?: string;
  error?: string;
  created: number;
  updated: number;
  skipped: number;
  failed: number;
  canUndo: boolean;
  createdByUserId?: string;
  createdByDisplayName?: string;
  createdAt: string;
  committedAt?: string;
  expiresAt: string;
}

export interface ImportSessionState {
  session: ImportSessionViewModel;
  sourceProfile?: ImportSourceProfile;
  mapping?: ImportMappingModel;
  previewResult?: ImportPreviewResult;
}

export interface ImportSourceColumn {
  index: number;
  name: string;
  inferredType: number;
  nonEmptyCount: number;
  distinctCount: number;
  sampleValues: string[];
}

export interface ImportSourceProfile {
  kind: number;
  encoding?: string;
  delimiter?: string;
  hasHeaderRow: boolean;
  sheetNames: string[];
  selectedSheet?: string;
  vendorProfile?: string;
  estimatedRowCount: number;
  columns: ImportSourceColumn[];
}

export interface ImportFieldBinding {
  fieldKey: string;
  columnIndex?: number | null;
  constant?: string | null;
  transforms: { kind: number; argument?: string }[];
  valueMap: Record<string, string>;
  confidence: number;
  origin: number;
}

export interface ImportMappingModel {
  recordType: string;
  bindings: ImportFieldBinding[];
  dedupe?: { keyFieldKey: string; action: number } | null;
}

export interface ImportRowDiagnostic {
  rowNumber: number;
  columnName?: string;
  severity: ImportDiagnosticSeverity;
  code: string;
  message: string;
  value?: string;
}

export interface ImportRowPreview {
  rowNumber: number;
  action: ImportRowAction;
  matchedRef?: string;
  resolved: Record<string, string | null>;
}

export interface ImportPreviewResult {
  totalRows: number;
  willCreate: number;
  willUpdate: number;
  willSkip: number;
  willError: number;
  isExtrapolated: boolean;
  diagnostics: ImportRowDiagnostic[];
  newEntities: { entityType: string; name: string }[];
  usersToInvite: string[];
  sampleRows: ImportRowPreview[];
}

export interface ImportMappingSuggestion {
  mapping: ImportMappingModel;
  vendor: number;
  vendorConfidence: number;
  unmappedColumns: number[];
}

export interface ImproveImportMappingResult {
  mapping: ImportMappingModel;
  discardedBindings: number;
  discardReasons: string[];
  notes?: string;
  usedDataSampling: boolean;
}

export interface ImportUndoResult {
  reverted: number;
  blocked: string[];
}

File: server/Netptune.Transfer/Entities/ImportSession.cs
using System.Text.Json;

using Netptune.Core.BaseEntities;
using Netptune.Transfer.Enums;

namespace Netptune.Transfer.Entities;

public record ImportSession : WorkspaceEntity<int>
{
    public const int DefaultRetentionDays = 7;

    public Guid PublicId { get; set; } = Guid.NewGuid();

    public ImportStage Stage { get; set; } = ImportStage.Uploaded;

    public ImportSourceKind SourceKind { get; set; }

    public ImportVendorProfile VendorProfile { get; set; } = ImportVendorProfile.None;

    public string OriginalName { get; set; } = null!;

    public string StorageKey { get; set; } = null!;

    public long SizeBytes { get; set; }

    public JsonDocument? SourceProfile { get; set; }

    public JsonDocument? Mapping { get; set; }

    public JsonDocument? PreviewResult { get; set; }

    public JsonDocument? Result { get; set; }

    public string TargetRecordType { get; set; } = null!;

    public string? TargetProjectKey { get; set; }

    public string? TargetBoardIdentifier { get; set; }

    public string CreatedBy { get; set; } = null!;

    public int ProgressPercent { get; set; }

    public string? ProgressMessage { get; set; }

    public string? Error { get; set; }

    public bool QuotaReleased { get; set; }

    public DateTime ExpiresAt { get; set; }

    public DateTime? CommittedAt { get; set; }

    public ICollection<ImportSessionEntry> Entries { get; set; } = new HashSet<ImportSessionEntry>();
}

public record ImportSessionEntry : KeyedEntity<long>
{
    public int SessionId { get; set; }

    public string EntityType { get; set; } = null!;

    public int EntityId { get; set; }

    public ImportEntryOperation Operation { get; set; }

    public JsonDocument? PreviousValues { get; set; }

    public DateTime? EntityUpdatedAt { get; set; }

    public ImportSession Session { get; set; } = null!;
}
